comm/src/main/scala/coop/rchain/comm/transport/StreamObservable.scala

```diff
@@ -11,38 +11,41 @@ import monix.reactive.Observable
 import monix.reactive.observers.Subscriber
 import java.nio.file._
 import scala.concurrent.duration._
+import coop.rchain.catscontrib.Catscontrib._
 
-case class ToStream(peerNode: PeerNode, path: Path, sender: PeerNode)
+case class StreamToPeers(peers: Seq[PeerNode], path: Path, sender: PeerNode)
 
 class StreamObservable(bufferSize: Int, folder: Path)(implicit log: Log[Task], scheduler: Scheduler)
-    extends Observable[ToStream] {
+    extends Observable[StreamToPeers] {
 
-  private val subject = buffer.LimitedBufferObservable.dropNew[ToStream](bufferSize)
+  private val subject = buffer.LimitedBufferObservable.dropNew[StreamToPeers](bufferSize)
 
   def stream(peers: List[PeerNode], blob: Blob): Task[Unit] = {
-    def push(peer: PeerNode): Task[Boolean] =
+
+    val storeBlob: Task[Option[Path]] =
       blob.packet.store[Task](folder) >>= {
-        case Right(file) => Task.delay(subject.pushNext(ToStream(peer, file, blob.sender)))
+        case Right(file) => Task.pure(Some(file))
         case Left(UnableToStorePacket(p, er)) =>
-          log.error(s"Could not serialize packet $p. Error message: $er") *> true.pure[Task]
+          log.error(s"Could not serialize packet $p. Error message: $er") *> None.pure[Task]
         case Left(er) =>
-          log.error(s"Could not serialize packet ${blob.packet}. Error: $er") *> true.pure[Task]
+          log.error(s"Could not serialize packet ${blob.packet}. Error: $er") *> None.pure[Task]
       }
 
-    def retry(failed: List[PeerNode]): Task[Unit] =
+    def push(file: Path): Task[Boolean] =
+      Task.delay(subject.pushNext(StreamToPeers(peers, file, blob.sender)))
+
+    def propose(file: Path): Task[Unit] =
+      push(file) >>= (_.fold(Task.unit, retry(file)))
```

Comment:
I know this is unrelated to this PR, but isn't that a potential infinite loop? we need to stop at some point?

Reply:
Yes, I had the same concern. On the other hand we are pushing an item to a queue. Eventually the queue will accept the item. Until then Casper must be blocked and not produce new items. We have limited reties (ATM 3x) on the other end of the message queue. So `propose`/`retry` should work as expected and not run forever.